### DISCUSSION TITLE
Thola support to query network devices via SNMP

### DISCUSSION THREAD

**Discussion by nick4me** [nautobot/nautobot#279] [Ideas]
it is very practical and useful that napalm is integrated into nautobot.
however, many of (our) existing network devices can only or at least easier be queried via SNMP.

here it would be great if thola were also integrated in addition to napalm:
https://github.com/inexio/thola
Thola supports a very large number of network devices and can be easily integrated via REST API.
Thola can also automatically identify a network device and determine the manufacturer, model and serial number.

**Comment by jathanism** [comment]
Hello @nick4me and thanks for the suggestion. I think this would be a good candidate for a discussion, because I think the idea is interesting.

I would like to add, however, that Thola is probably not a good fit for direct integration since it is not a Python project (it's written in Go), so it would have to be external. So perhaps a plugin that would implement a Thola client in Python that could integrate with an instance of Thola might be feasible. 

If you're agreeable, I'll convert this issue to a discussion and see where that takes us?

**Comment by nick4me** [comment]
@jathanism you are right. if you like the idea we could also write a lightweight python client library for Thola. maybe you can specify the requirements (methods with input and response parameters, ...) and we will provide such a library in the next few days.

I am very happy to discuss the integration. As user of nautobot we would be very happy to integrate the realtime queries and the onboarding process for network devices with Thola.

**Reply by mzbroch** [reply]
I find it an interesting idea, although Onboarding Plugin has a tight integration with Napalm which also spans to onboarding's extensions. (customer's code to execute internal logic on the napalm)

Onboarding also implements platform auto-discovery and this one relies on the Netmiko device types.

**Reply by nick4me** [reply]
we have updated https://github.com/inexio/thola-ansible and added some more examples.
also for identifying network devices with ansible and feeding back collected facts to nautobot.
For example, the operating system version determined by thola or the network interfaces read out can be played back to nautobot. this is also a functionality that helps with onboarding.

**Comment by nick4me** [comment]
The thola client library for python has been uploaded:
https://github.com/inexio/thola-client-module-python

There is also a module for ansible (specially to gather facts of network devices):
https://github.com/inexio/thola-ansible
